crates/accelerate/src/high_level_synthesis.rs

```diff
@@ -171,82 +213,767 @@ impl QubitTracker {
+                        }
+                    }
+                }
+
+                // Add the operations from the circuit synthesized for the current operation to the output circuit.
+                // The correspondence between qubits is:
+                // qubit index in the synthesized circuit -> corresponding global qubit -> corresponding qubit in the output circuit
+                let qubit_map: HashMap<usize, usize> = HashMap::from_iter(
+                    synthesized_circuit_qubits
+                        .iter()
+                        .enumerate()
+                        .map(|(i, q)| (i, global_to_local[q])),
+                );
+
+                for inst_inner in synthesized_circuit.iter() {
+                    let inst_inner_qubits = synthesized_circuit.get_qargs(inst_inner.qubits);
+                    let inst_inner_clbits = synthesized_circuit.get_cargs(inst_inner.clbits);
+
+                    let inst_outer_qubits: Vec<Qubit> = inst_inner_qubits
+                        .iter()
+                        .map(|q| Qubit::new(qubit_map[&q.index()]))
+                        .collect();
+                    let inst_outer_clbits: Vec<Clbit> =
+                        inst_inner_clbits.iter().map(|c| Clbit(c.0)).collect();
+
+                    output_circuit.push_packed_operation(
+                        inst_inner.op.clone(),
+                        inst_inner.params_view(),
+                        &inst_outer_qubits,
+                        &inst_outer_clbits,
+                    )?;
+                }
+
+                let updated_global_phase = radd_param(
+                    output_circuit.global_phase().clone(),
+                    synthesized_circuit.global_phase().clone(),
+                    py,
+                );
+                output_circuit.set_global_phase(py, updated_global_phase)?;
+            }
+        }
+    }
+
+    // Another pedantic check that can possibly be removed.
+    if output_circuit.num_qubits() != output_qubits.len() {
+        return Err(TranspilerError::new_err(
+            "HighLevelSynthesis: the output from 'run_on_circuitdata' is incorrect.",
+        ));
+    }
+
+    Ok((output_circuit, output_qubits))
+}
+
+/// Produces a definition circuit for an operation.
+///
+/// Essentially this function constructs a default definition for a unitary gate, in which case
+/// ``op.definition`` purposefully returns ``None``.
+/// For all other operation types, it simply calls ``op.definition``.
+fn extract_definition(
+    py: Python,
+    op: &PackedOperation,
+    params: &[Param],
+) -> PyResult<Option<CircuitData>> {
+    match op.view() {
+        OperationRef::Unitary(unitary) => {
+            let unitary: Array<Complex<f64>, Dim<[usize; 2]>> = match unitary.matrix(&[]) {
+                Some(unitary) => unitary,
+                None => return Err(TranspilerError::new_err("Unitary not found")),
+            };
+            match unitary.shape() {
```

Comment:
Isn't this something that should be taken care of by unitary synthesis at the moment? Otherwise we're duplicating the logic and run the risk of hardcoding the target basis here, whereas the unitary synthesis can handle more general cases, no?

Reply:
I agree, I would have also liked to use some readily available unitary synthesis code :).